drivers/SmartThings/jbl/src/lunchbox/sse/eventsource.lua

```diff
@@ -0,0 +1,500 @@
+        return nil
+      else
+        remainder_sent = true
+        return remaining, false
+      end
+    end
+    local next_line = string.sub(remaining, 1, line_end - 1)
+    remaining = string.sub(remaining, rn_end + 1)
+    return next_line, true
+  end
+end
+--- SSE spec for interpreting an event stream:
+--- https://html.spec.whatwg.org/multipage/server-sent-events.html#the-eventsource-interface
+--- @function parse
+--- @local
+--- @param source EventSource
+--- @param recv string the received payload from the last socket receive
+local function sse_parse_chunk(source, recv)
+  for line, complete in event_lines(recv) do
+    if not source._needs_more and (#line == 0 or (not line:match("([%w%p]+)"))) then -- empty/blank lines indicate dispatch
+      dispatch_event(source)
+    elseif source._needs_more then
+      local append = line
+      if source._last_field == "data" and complete then append = append .. "\n" end
+      if complete then source._needs_more = false end
+      source._parse_buffers[source._last_field] = source._parse_buffers[source._last_field] .. append
+    else
+      if line:sub(1, 1) ~= ":" then                  -- ignore any complete lines that start w/ a colon
+        local matches = line:gmatch("(%w*)(:*)(.*)") -- colon after field is optional, in that case it's a field w/ no value
+
+        for field, _colon, value in matches do
+          value = value:gsub("^[^%g]", "", 1) -- trim a single leading space character
+
+          if valid_fields[field] then
+            source._last_field = field
+            if field == "retry" then
+              local new_time = tonumber(value, 10)
+              if type(new_time) == "number" then
+                source._reconnect_time_millis = new_time
+              end
+            elseif field == "data" then
+              local append = (value or "")
+              if complete then append = append .. "\n" end
+              source._parse_buffers[field] = source._parse_buffers[field] .. append
+            elseif field == "id" then
+              -- skip ID's if they contain the NULL character
+              if not string.find(value, '\0') then
+                source._parse_buffers[field] = value
+              end
+            else
+              source._parse_buffers[field] = value
+            end
+          end
+          source._needs_more = source._needs_more or (not complete)
+        end
+      end
+    end
+  end
+end
+
+--- Helper function that captures the cyclic logic of the EventSource while in the CONNECTING state.
+--- @function connecting_action
+--- @local
+--- @param source EventSource
+local function connecting_action(source)
+  if not source._sock then
+    if type(source._sock_builder) == "function" then
+      source._sock = source._sock_builder()
+    else
+      source._sock, err = socket.tcp()
+      if err ~= nil then return nil, err end
+
+      _, err = source._sock:connect(source.url.host, source.url.port)
```

Comment:
I would recommend setting a timeout on this socket. We made a similar change in our versions of Lunchbox and added logic to the state machine processing loop to not treat timeouts as fatal errors.

Reply:
update this file to latest version in sample